File: leetcode/easy/count-special-quadruplets.ts
// https://leetcode.com/problems/count-special-quadruplets/

// MINE: brute force
/*
function countQuadruplets(nums: number[]): number {
    let n = nums.length, res = 0;
    for (let i = 0; i < n; i++) {
        for (let j = i + 1; j < n; j++) {
            for (let k = j + 1; k < n; k++) {
                for (let m = k + 1; m < n; m++) {
                    if (nums[i] + nums[j] + nums[k] == nums[m]) {
                        res++;
                    }
                }
            }
        }
    }
    return res;
};
*/

// ANSWER: O(3)
function countQuadruplets(nums: number[]): number {
  const seen = new Map()
  const n = nums.length
  let count = 0
  for (let i = n - 1; i >= 0; --i) {
    for (let j = i - 1; j >= 0; --j) {
      for (let k = j - 1; k >= 0; --k) {
        const possibleSum = nums[i] + nums[j] + nums[k]
        if (seen.has(possibleSum)) count += seen.get(possibleSum)
      }
    }
    seen.set(nums[i], seen.get(nums[i]) + 1 || 1)
  }
  return count
}

File: leetcode/easy/count-special-quadruplets.py
# https://leetcode.com/problems/count-special-quadruplets/

class Solution:
    def countQuadruplets(self, nums: List[int]) -> int:
        seen = defaultdict(int)
        n = len(nums)
        count = 0
        for i in range(n-1, -1, -1):
            for j in range(i-1, -1, -1):
                for k in range(j-1, -1, -1):
                    possibleSum = nums[i] + nums[j] + nums[k]
                    if possibleSum in seen:
                        count += seen[possibleSum]
            seen[nums[i]] += 1
        return count
